### ISSUE TITLE
[iOS] Some text is displaying incorrectly on iOS 14

### ISSUE THREAD

**Issue by EliasDeuss** [flutter/flutter#60013]
Looks like some text widgets display incorrectly in iOS 14 Beta 1. Looks like flutter is squishing some of the text.

[iOS 13.6]
![image](https://user-images.githubusercontent.com/13773566/85334247-c2381600-b48f-11ea-88e2-029725d4199f.jpeg)
![IMG_2775](https://user-images.githubusercontent.com/13773566/85827065-4ca3a280-b73a-11ea-9221-0824b2b36380.jpeg)

[iOS 14 Beta 1]
![image](https://user-images.githubusercontent.com/13773566/85334282-d419b900-b48f-11ea-9aa9-60228a2930f3.jpeg)
![IMG_0232](https://user-images.githubusercontent.com/13773566/85826999-254cd580-b73a-11ea-945d-e8f59adb1e17.jpeg)

**Comment by xster** [comment]
@GaryQian might have some thoughts. Could be a skia issue.

**Comment by gaaclarke** [comment]
What font is this using?  It could just be the font got updated if it's a system font?

**Comment by xster** [comment]
can confirm, the fonts do look pretty squished 

<img width="487" alt="Screen Shot 2020-06-22 at 4 29 27 PM" src="https://user-images.githubusercontent.com/156888/85344811-8c9e2780-b4a5-11ea-8c6a-4eac36c6e467.png">

**Comment by EliasDeuss** [comment]
> What font is this using? It could just be the font got updated if it's a system font?

I believe it is only system fonts that are effected. @gaaclarke @xster

**Comment by NoahJApps** [comment]
> > What font is this using? It could just be the font got updated if it's a system font?
> 
> I believe it is only system fonts that are effected. @gaaclarke @xster

I can confirm this from what I've investigated

**Comment by GaryQian** [comment]
We do not apply any x-axis scaling transforms to font glyphs, but in this it looks like the glyphs are actually narrower than they are supposed to be. This must mean that somehow, skia is interpreting the font glyphs differently. It could just mean that skia needs slight changes to support iOS 14.

@Rusino (will link this bug to additional skia folks)

**Comment by GaryQian** [comment]
Relaying Ben Wagner on the Skia team's initial response:

> That font looks like the system hidden San Francisco font. This font has been a variable font now for a bit, and it's entirely possible that the default instance was made narrower. Are you sure this font isn't just narrower? The other thing that can happen is that now that it's a system font they may be exposing more instances so you're now getting something closer to what you actually requested (many were surprised then their request for '300' weight actually started producing '300' weight instead of '400' weight).
> 
> This might happen if there were only a weight variation, but now both the weight and width variations are together in the one font and now you actually get everything you ask for.

**Comment by EliasDeuss** [comment]
@xster It looks like it also affects pop up dialogues and copy and past prompts.

![image](https://user-images.githubusercontent.com/13773566/85963970-b0adad00-b96c-11ea-909d-7a5874d281fd.jpeg)

**Comment by chinmaygarde** [comment]
xref b/160564432

**Comment by maxbeech** [comment]
I have found this issue vis a Google Font too (Red Hat Text). Any ideas on a fix please?

**Comment by xster** [comment]
![Screen Shot 2020-08-11 at 11 42 00 PM](https://user-images.githubusercontent.com/156888/89983372-486d0f00-dc2c-11ea-866d-a2232b2ba4b3.png)

It's not the system font itself. Comparing in iOS 13 and 14 vs the system font via Xcode looks the same. But Flutter's rendering is off on 14.

**Comment by xster** [comment]
Trying to document discoveries. Turns out it's slightly more complicated than I thought but solvable. 

Ben thinks the issue is the increased complexity in the variability of the fonts. e.g.  

```
Beginning in iOS 14, the system provides the San Francisco and New York fonts in the variable font format. This format combines different font styles together in one file, and supports interpolation between styles to create intermediate ones. With interpolation, typefaces can adapt to all sizes while appearing specifically designed for each size.

Interpolation also enables optical sizing, which refers to the creation of different typographic designs to fit different sizes. Both San Francisco and New York provide specific optical size variants to ensure that text can look great at any size: Text and Display for SF Pro and SF Compact, and Small, Medium, Large, and Extra Large for New York. In iOS 14 and later, the system fonts support dynamic optical sizes, merging the discrete optical sizes like Text and Display into a single, continuous design. This design allows each glyph or letterform to be interpolated to produce a structure that's precisely adapted to the point size.
```

from https://developer.apple.com/design/human-interface-guidelines/ios/visual-design/typography/. 

In other words, the SkFontMgr that we've been delegating to won't be able to resolve to the specific font in a way that matches as tightly to the system as before. 

He still recommends that we confirm this issue by looking for the CTFontRef used to create the SkTypeface (with SkTypeface_GetCTFontRef ) and CFShow it to see if we're indeed resolving to something different than plain Xcode showing a .SFUI-Regular in a UILabel. 

In this case, we'll need to resolve a CTFont ourselves to create a SkTypeface instead of using the SkFontMgr (https://github.com/flutter/engine/blob/9f146b5f471ba9c3f9088ac9b19fc758a6acdc55/third_party/txt/src/txt/font_collection.cc#L211). Chrome currently does https://source.chromium.org/chromium/chromium/src/+/master:ui/gfx/platform_font_ios.mm;l=103;drc=3930cf3f0e2404a517b4580b21e598a3fc648ff0. 

https://developer.apple.com/videos/play/wwdc2019/227/ is the background. 

@jason-simmons recommends adding another SkFontMgr implementation that fulfills this on iOS in this fallback chain https://github.com/flutter/engine/blob/master/third_party/txt/src/txt/font_collection.cc#L112. 

We'll also need to migrate this workaround to SkParagraph https://github.com/google/skia/blob/master/modules/skparagraph/src/FontCollection.cpp. 

There's also the matter of non-SF system fonts being different but Ben thinks it's just a kerning issue. 
![Screen Shot 2020-08-12 at 11 20 52 AM](https://user-images.githubusercontent.com/156888/90074383-f1a91900-dcaf-11ea-823f-208555303291.png)

**Comment by xster** [comment]
I created a version of the SkFontMgr that wraps the SkFontMgr_Mac to do non font-family-name-string based font look-ups but then realized that the problem in this case is lamer than previously thought. 

The above post still applies and it's a systemic risk that we still have system font lookups based on name strings (see WWDC slides https://devstreaming-cdn.apple.com/videos/wwdc/2019/227a5yni0swztyj0/227/227_font_management_and_text_scaling.pdf?dl=1)

![Screen Shot 2020-08-17 at 9 58 45 AM](https://user-images.githubusercontent.com/156888/90422859-524da280-e070-11ea-877c-99b7c755bcad.png)

But at the present, it still works. It's still risky since most lookups by name will fail whether you look for ".SF Pro Text" or "".SFUI-Regular" or some such. We're just lucky that we almost always lookup system fonts (fonts starting with dots) for SF fonts and our fallback font (https://cs.opensource.google/flutter/engine/+/master:third_party/txt/src/txt/platform_mac.mm;l=21;drc=8ff6948e23c20e1c9382429669721da4a1998a1b) is the SF font. The internal string is family name is ".AppleSystemUIFont" and that lookup by `CTFontDescriptorCreateWithAttributes` via `kCTFontFamilyNameAttribute` still works and [normalizing via `CTFontDescriptorCreateMatchingFontDescriptors`](https://cs.opensource.google/skia/skia/+/master:src/ports/SkFontMgr_mac_ct.cpp;l=345;drc=0ef90dd5e9c6526530f14f6943f304515140165a) against `kCTFontFamilyNameAttribute` still produces valid CTFontRefs. Except it makes more width `kCTFontWidthTrait` variants now from the same attributes lookup. 

Since Flutter's TextStyle doesn't have a width API (since it's typically directly specified in the family name such as `.SFUI-SemiCondensedUltralight`), the [default sorting](https://cs.opensource.google/flutter/engine/+/master:third_party/txt/src/txt/font_collection.cc;l=230;drc=e55b4476f1657590783ca04865d7b354fa4559dc) ends up (incorrectly) determining the width we use.

**Comment by EliasDeuss** [comment]
@xster When will this reach the Stable branch?

**Comment by xster** [comment]
In 2-3 weeks I'm guessing

**Comment by ltdangkhoa** [comment]
@xster Sorry to remind this issue. I've a problem with publishing app to App Store related to font issues:

> ITMS-90853: Invalid font - There was an error validating the font named 'All3PIcons-Regular' at 'Runner.app/Frameworks/App.framework/flutter_assets/fonts/MaterialIcons-Regular.otf'. ‘name’ table structure.
> 
> ITMS-90853: Invalid font - There was an error validating the font named 'All3PIcons-Regular' at 'Runner.app/Frameworks/App.framework/flutter_assets/fonts/MaterialIcons-Regular.otf'. ‘name’ table usability.

Do you have any idea about this problem?

**Comment by xster** [comment]
Can you open a new issue? We can ask the material icons folks.

**Comment by miguelpruivo** [comment]
When is this expected to land on stable?

**Comment by tobinharris** [comment]
I'm on Flutter 1.20.4 and guessing I'll need to wait until 1.22.x in order to see a fix for this to appear?

**Comment by jmagman** [comment]
This was merged in #64525, which is available in 1.22.0-9.0.pre, and is currently available on the [master, dev, and beta channels](https://flutter.dev/docs/development/tools/sdk/releases?tab=macos).

**Comment by FickleLife** [comment]
Thanks @jmagman - any idea how long it will be to hit stable branch?

**Comment by xster** [comment]
More than 1 day, less than 1 month.

**Comment by lucien144** [comment]
I can confirm this has been resolved in Flutter 1.22.0-12.3.pre
